plugins/experimental/ja4_fingerprint/ja4.h

```diff
@@ -0,0 +1,156 @@
+  limitations under the License.
+
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <iterator>
+#include <string>
+#include <vector>
+
+namespace JA4
+{
+
+constexpr char delimiter{'_'};
+
+enum class Protocol {
+  DTLS = 'd',
+  QUIC = 'q',
+  TLS  = 't',
+};
+
+enum class SNI {
+  to_domain = 'd',
+  to_IP     = 'i',
+};
+
+/**
+ * Represents the data sent in a TLS Client Hello needed for JA4 fingerprints.
+ */
+class TLSSummary
+{
+public:
+  using difference_type = std::iterator_traits<std::vector<std::uint16_t>::iterator>::difference_type;
+
+  Protocol    protocol;
+  SNI         SNI_type;
+  std::string TLS_version;
+  std::string ALPN;
+
+  std::vector<std::uint16_t> const &get_ciphers() const;
+  void                              add_cipher(std::uint16_t cipher);
+
+  std::vector<std::uint16_t> const &get_extensions() const;
+  void                              add_extension(std::uint16_t extension);
+
+  /**
+   * Get the number of ciphers excluding GREASE values.
+   *
+   * @return Returns the count of non-GREASE ciphers.
+   */
+  difference_type get_cipher_count() const;
+
+  /**
+   * Get the number of extensions excluding GREASE values.
+   *
+   * @return Returns the count of non-GREASE extensions.
+   */
+  difference_type get_extension_count() const;
+
+private:
+  std::vector<std::uint16_t> _ciphers;
+  std::vector<std::uint16_t> _extensions;
+  int                        _extension_count_including_sni_and_alpn;
+};
+
+/**
+ * Calculate the a portion of the JA4 fingerprint for the given TLS summary.
+ *
+ * The a portion of the fingerprint encodes the protocol, TLS version, SNI
+ * type, number of cipher suites, number of extensions, and first ALPN value.
+ *
+ * For more information see
+ * https://github.com/FoxIO-LLC/ja4/blob/main-technical_details/JA4.md.
+ *
+ * @param TLS_summary The TLS client hello.
+ * @return Returns a string containing the a portion of the JA4 fingerprint.
+ */
+std::string make_JA4_a_raw(TLSSummary const &TLS_summary);
+
+/**
+ * Calculate the b portion of the JA4 fingerprint for the given TLS summary.
+ *
+ * The b portion of the fingerprint is a comma-delimited list of lowercase hex
+ * numbers representing the cipher suites in sorted order. GREASE values are
+ * ignored.
+ *
+ * For more information see
+ * https://github.com/FoxIO-LLC/ja4/blob/main-technical_details/JA4.md.
+ *
+ * @param TLS_summary The TLS client hello.
+ * @return Returns a string containing the b portion of the JA4 fingerprint.
+ */
+std::string make_JA4_b_raw(TLSSummary const &TLS_summary);
+
+/**
+ * Calculate the c portion of the JA4 fingerprint for the given TLS summary.
+ *
+ * The b portion of the fingerprint is a comma-delimited list of lowercase hex
+ * numbers representing the extensions in sorted order. GREASE values and the
+ * SNI and ALPN extensions are ignored.
+ *
+ * For more information see
+ * https://github.com/FoxIO-LLC/ja4/blob/main-technical_details/JA4.md.
```

Comment:
These look like a dead links to me. Looking into it a bit, I think `main-technical` should be `main/technical`